File: wind-composer-web/src/music/arrangementEngine.ts
import { clamp } from "../utils";
import type { StyleProfile } from "./styleEngine";

export const SONG_SECTIONS = [
  "Intro",
  "Verse",
  "Build",
  "Drop",
  "Break",
  "Recovery",
  "Outro",
  "Flow",
] as const;

export type SongSectionName = typeof SONG_SECTIONS[number];

export interface ArrangementState {
  section: SongSectionName;
  barsInSection: number;
  sectionEnergy: number;
  activeLayers: Set<string>;
}

export class ArrangementEngine {
  private state: ArrangementState = {
    section: "Flow",
    barsInSection: 0,
    sectionEnergy: 0.5,
    activeLayers: new Set(["main_pad", "atmosphere"]),
  };
  private sectionIdx = 0;

  getState(): ArrangementState {
    return this.state;
  }

  /** Call once per bar (measure), not every composition tick. */
  onBar(measure: number, energy: number, storm: boolean): SongSectionName {
    if (measure <= 0) {
      this.state.sectionEnergy = this.sectionEnergy(energy);
      return this.state.section;
    }

    this.state.barsInSection += 1;
    if (this.state.barsInSection >= 32 || (storm && this.state.section !== "Drop")) {
      this.sectionIdx = (this.sectionIdx + 1) % SONG_SECTIONS.length;
      if (storm && Math.random() < 0.55) {
        this.state.section = "Drop";
      } else {
        this.state.section = SONG_SECTIONS[this.sectionIdx];
      }
      this.state.barsInSection = 0;
    }
    this.state.sectionEnergy = this.sectionEnergy(energy);
    this.state.activeLayers = this.layersForSection(this.state.section);
    return this.state.section;
  }

  layerGains(style: StyleProfile, energy: number): Record<string, number> {
    const s = this.state.section;
    const base: Record<string, number> = {
      main_pad: style.padLayers,
      soft_bass: style.bassLayers * energy,
      sub_bass: style.bassLayers * energy * 0.82,
      lead: style.leadLayers * energy,
      atmosphere: style.padLayers * 0.48,
      percussion: style.drumDensity * energy,
    };
    if (s === "Intro") {
      base.lead *= 0.28;
      base.percussion *= 0.18;
    } else if (s === "Verse") {
      base.percussion *= 0.65;
      base.lead *= 0.55;
    } else if (s === "Build") {
      base.percussion *= 1.25;
      base.lead *= 0.68;
    } else if (s === "Drop") {
      base.percussion *= 1.45;
      base.lead *= 1.22;
      base.sub_bass *= 1.35;
    } else if (s === "Break") {
      base.percussion *= 0.12;
      base.lead *= 0.35;
    } else if (s === "Outro") {
      base.percussion *= 0.08;
      base.lead *= 0.18;
    }
    return Object.fromEntries(Object.entries(base).map(([k, v]) => [k, clamp(v)]));
  }

  private sectionEnergy(energy: number): number {
    const s = this.state.section;
    if (s === "Drop") return clamp(energy + 0.28);
    if (s === "Break") return clamp(energy * 0.48);
    if (s === "Build") return clamp(energy + 0.18);
    return clamp(energy);
  }

  private layersForSection(section: SongSectionName): Set<string> {
    const layers = new Set(["main_pad", "atmosphere"]);
    if (["Build", "Drop", "Flow", "Verse"].includes(section)) {
      layers.add("soft_bass");
      layers.add("percussion");
      layers.add("lead");
    }
    if (section === "Drop") layers.add("sub_bass");
    if (section === "Break") {
      layers.clear();
      layers.add("main_pad");
      layers.add("atmosphere");
      layers.add("choir");
    }
    return layers;
  }

  reset(): void {
    this.state = {
      section: "Flow",
      barsInSection: 0,
      sectionEnergy: 0.5,
      activeLayers: new Set(["main_pad", "atmosphere"]),
    };
    this.sectionIdx = 0;
  }
}

File: wind-composer/arrangement_engine.py
"""Song sections and layer arrangement."""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import Dict, Set

from style_engine import SongSection, StyleProfile
from utils import clamp


@dataclass
class ArrangementState:
    section: SongSection = SongSection.FLOW
    bars_in_section: int = 0
    active_layers: Set[str] = field(default_factory=lambda: {"main_pad", "atmosphere"})
    section_energy: float = 0.5


class ArrangementEngine:
  SECTION_CYCLE = [
    SongSection.INTRO,
    SongSection.BUILD,
    SongSection.DROP,
    SongSection.BREAKDOWN,
    SongSection.RECOVERY,
    SongSection.FLOW,
  ]

  def __init__(self) -> None:
    self._state = ArrangementState()
    self._section_idx = 0

  @property
  def state(self) -> ArrangementState:
    return self._state

  def on_bar(self, measure: int, energy: float, storm: bool) -> SongSection:
    """Advance arrangement once per bar, not every composition tick."""
    if measure <= 0:
      self._state.section_energy = self._section_energy(energy)
      return self._state.section

    self._state.bars_in_section += 1
    if self._state.bars_in_section >= 32 or (storm and self._state.section != SongSection.DROP):
      self._section_idx = (self._section_idx + 1) % len(self.SECTION_CYCLE)
      if storm and random.random() < 0.6:
        self._state.section = SongSection.DROP
      else:
        self._state.section = self.SECTION_CYCLE[self._section_idx]
      self._state.bars_in_section = 0
    self._state.section_energy = self._section_energy(energy)
    self._state.active_layers = self._layers_for_section(self._state.section)
    return self._state.section

  def on_phrase(self, energy: float, storm: bool) -> SongSection:
    """Legacy alias — prefer on_bar with measure index."""
    return self.on_bar(1, energy, storm)

  def layer_gains(self, style: StyleProfile, energy: float) -> Dict[str, float]:
    s = self._state.section
    base = {
      "main_pad": style.pad_layers,
      "soft_bass": style.bass_layers * energy,
      "sub_bass": style.bass_layers * energy * 0.8,
      "lead": style.lead_layers * energy,
      "atmosphere": style.pad_layers * 0.5,
      "percussion": style.drum_density * energy,
    }
    if s == SongSection.INTRO:
      base["lead"] *= 0.3
      base["percussion"] *= 0.2
    elif s == SongSection.BUILD:
      base["percussion"] *= 1.2
      base["lead"] *= 0.7
    elif s == SongSection.DROP:
      base["percussion"] *= 1.4
      base["lead"] *= 1.2
      base["sub_bass"] *= 1.3
    elif s == SongSection.BREAKDOWN:
      base["percussion"] *= 0.15
      base["lead"] *= 0.4
    elif s == SongSection.OUTRO:
      base["percussion"] *= 0.1
      base["lead"] *= 0.2
    return {k: clamp(v) for k, v in base.items()}

  def _section_energy(self, energy: float) -> float:
    s = self._state.section
    if s == SongSection.DROP:
      return clamp(energy + 0.25)
    if s == SongSection.BREAKDOWN:
      return clamp(energy * 0.5)
    if s == SongSection.BUILD:
      return clamp(energy + 0.15)
    return clamp(energy)

  def _layers_for_section(self, section: SongSection) -> Set[str]:
    layers = {"main_pad", "atmosphere"}
    if section in (SongSection.BUILD, SongSection.DROP, SongSection.FLOW):
      layers |= {"soft_bass", "percussion", "lead"}
    if section == SongSection.DROP:
      layers |= {"sub_bass"}
    if section == SongSection.BREAKDOWN:
      layers = {"main_pad", "atmosphere", "choir"}
    return layers
